demo/actions.py

```diff
@@ -17,35 +17,172 @@
+                                                           domain))
+
+            if not slot_values:
+
+                # if no email entity was picked up, but intent was
+                # enter_data, ask again instead of leaving the form
+                intent = tracker.latest_message['intent'].get('name')
+                if intent == "enter_data":
+                    dispatcher.utter_template('utter_no_email', tracker)
+                    return []
+
+                # reject to execute the form action
+                # if some slot was requested but nothing was extracted
+                # it will allow other policies to predict another action
+                raise ActionExecutionRejection(self.name(),
+                                               "Failed to extract slot {0} "
+                                               "with action {1}"
+                                               "".format(slot_to_fill,
+                                                         self.name()))
+
+            for slot, value in slot_values.items():
+                validate_func = getattr(self, "validate_{}".format(slot),
+                                        lambda *x: value)
+                slot_values[slot] = validate_func(value, dispatcher, tracker,
+                                                  domain)
+
+        # validation succeed, set slots to extracted values
+        return [SlotSet(slot, value) for slot, value in slot_values.items()]
+
+    def submit(self,
+               dispatcher: CollectingDispatcher,
+               tracker: Tracker,
+               domain: Dict[Text, Any]) -> List[Dict]:
+        """Once we have an email, attempt to add it to the database"""
 
-    def run(self, dispatcher, tracker, domain):
         email = tracker.get_slot('email')
-        if email:
-            client = MailChimpAPI(config.mailchimp_api_key)
-            # if the email is already subscribed, this returns False
-            subscribed = client.subscribe_user(config.mailchimp_list, email)
+        client = MailChimpAPI(config.mailchimp_api_key)
+        # if the email is already subscribed, this returns False
+        added_to_list = client.subscribe_user(config.mailchimp_list, email)
 
-            return [SlotSet('subscribed', subscribed)]
+        # utter submit template
+        if added_to_list:
+            dispatcher.utter_template('utter_confirmationemail', tracker)
+        else:
+            dispatcher.utter_template('utter_already_subscribed', tracker)
         return []
 
 
-class ActionStoreSalesInfo(Action):
-    """Saves the information collected in the sales flow into a spreadsheet"""
+class SalesForm(FormAction):
+    """Collects sales information and adds it to the spreadsheet"""
 
     def name(self):
-        return "action_store_sales_info"
+        return "sales_form"
+
+    @staticmethod
+    def required_slots(tracker):
+        return ["job_function", "use_case", "budget", "person_name",
```

Comment:
ok i remember why we had separate slot names/entities. I think it was because the entity wasn't always picked up and so we wanted to make sure the right thing was stored. I think it's ok to make the slot/entity names matchup now, apart from business_email and budget. 

As for business_email, I actually think it would be cool to check if the value of the email slot should be used as their business email. I.e. if people already provided it for the newsletter, they may want to reuse it again

Reply:
ooh i like that idea!